#### Setup

In [1]:
import polars as pl
from analysis.characterisation.clustering import cluster_timeseries_usage, usage_probabilities
from analysis.characterisation.notebooks.notebook_config import (
    dl,
    FEATURES,
    N_CLUSTERS,
    DATASET_START,
    DATASET_END,
    TIME_SERIES_MODE,
    WINDOW_MONTHS,
)

usage = cluster_timeseries_usage(
    loader=dl,
    k=N_CLUSTERS,
    features=FEATURES,
    start=DATASET_START,
    end=DATASET_END,
    mode=TIME_SERIES_MODE,
    window_months=WINDOW_MONTHS
)

usage_probs = usage_probabilities(usage).sort(["station", "probability"], descending=True)


Perform Clustering in Interval 2016-01-01 until 2018-01-01
Perform Clustering in Interval 2016-02-01 until 2018-02-01
Perform Clustering in Interval 2016-03-01 until 2018-03-01
Perform Clustering in Interval 2016-04-01 until 2018-04-01
Perform Clustering in Interval 2016-05-01 until 2018-05-01
Perform Clustering in Interval 2016-06-01 until 2018-06-01
Perform Clustering in Interval 2016-07-01 until 2018-07-01
Perform Clustering in Interval 2016-08-01 until 2018-08-01
Perform Clustering in Interval 2016-09-01 until 2018-09-01
Perform Clustering in Interval 2016-10-01 until 2018-10-01
Perform Clustering in Interval 2016-11-01 until 2018-11-01
Perform Clustering in Interval 2016-12-01 until 2018-12-01
Perform Clustering in Interval 2017-01-01 until 2019-01-01
Perform Clustering in Interval 2017-02-01 until 2019-02-01
Perform Clustering in Interval 2017-03-01 until 2019-03-01
Perform Clustering in Interval 2017-04-01 until 2019-04-01
Perform Clustering in Interval 2017-05-01 until 2019-05-

## 7. Impact of Weather on Station Usage Patterns

### Event based effects

We will now investiage the effects of predefined weather effects on different usage types. 

Before we start we have to figure out how we should define the temperature, perticipation and wind classes. 

For this we will take a look at the weather data of Heidelberg.

#### Weather class definitions
Temperature classes were defined using empirical terciles (Q33, Q66).
For Precipitation zero values (dry days) were filtered out.

In [ ]:
wd = dl.get_weather().df

wd_daily = (
    wd
    .with_columns(pl.col("datetime").dt.date().alias("date"))
    .group_by("date")
    .agg(
        pl.col("temperature_2m").max().alias("temp_max"),
        pl.col("precipitation").sum().alias("precip_sum"),
        pl.col("wind_speed_10m").max().alias("wind_max"),
    )
)

temp_quantiles = (
    wd_daily
    .select(
        pl.col("temp_max").quantile(0.33).alias("q33"),
        pl.col("temp_max").quantile(0.66).alias("q66"),
    )
)

precip_quantiles = (
    wd_daily
    .filter(pl.col("precip_sum") > 0)
    .select(
        pl.col("precip_sum").quantile(0.8).alias("q80"),
    )
)

wind_quantiles = (
    wd_daily
    .select(
        pl.col("wind_max").quantile(0.33).alias("q33"),
        pl.col("wind_max").quantile(0.66).alias("q66"),
    )
)


print("Temperature quantiles (daily max)")
display(temp_quantiles)

print("\nPrecipitation quantiles (daily sum)")
display(precip_quantiles)

print("Wind quantiles (daily max)")
display(wind_quantiles)

Temperature quantiles


shape: (1, 2)
┌──────┬──────┐
│ q33  ┆ q66  │
│ ---  ┆ ---  │
│ f64  ┆ f64  │
╞══════╪══════╡
│ 10.7 ┆ 19.8 │
└──────┴──────┘


Precipitation quantiles ()


shape: (1, 1)
┌─────┐
│ q80 │
│ --- │
│ f64 │
╞═════╡
│ 7.1 │
└─────┘

Wind quantiles (daily max)


shape: (1, 2)
┌──────┬──────┐
│ q33  ┆ q66  │
│ ---  ┆ ---  │
│ f64  ┆ f64  │
╞══════╪══════╡
│ 12.0 ┆ 16.8 │
└──────┴──────┘

**Temperature**

We define three different temperature ranges:
$$
\begin{aligned}
\text{L (Low):} \quad & \max(T_{\text{day}}) < 10.7^\circ \mathrm{C} \\
\text{M (Medium):} \quad & 10.7 \leq \max(T_{\text{day}}) < 19.8^\circ \mathrm{C} \\
\text{H (High):} \quad & \max(T_{\text{day}}) \geq 19.8^\circ \mathrm{C}
\end{aligned}
$$

implemented in `temp_ranges`.

In [3]:
def temp_ranges(col):
    return (
        pl.when(col < 7.2).then(pl.lit("L"))
        .when(col < 15.1).then(pl.lit("M"))
        .otherwise(pl.lit("H"))
    )

We will use low temperatures (*L*) as baseline and calculate the relative change to range *M* and *H* per station. 

$$
\text{rel\_diff}_X = \frac{\bar{C}_X - \bar{C}_L}{\bar{C}_L} \times 100 
\quad \text{for } X \in \{M, H\}
$$

Furthermore we will look seperately at weekdays and weekends.

In [4]:
from analysis.characterisation.weather import weather_response_df
from analysis.characterisation.helpers import dominant_usage_per_station

def base_df(loader, usage_probs, sample_rate="1h"):
  df = weather_response_df(loader=loader, sample_rate=sample_rate)

  # add dominant usage type 
  df = df.join(dominant_usage_per_station(usage_probs), on="station", how="left")

  # add weekday / weekend classification
  df = df.with_columns([ 
    pl.when(pl.col("datetime").dt.weekday() <= 4)
      .then(pl.lit("weekday")) 
      .otherwise(pl.lit("weekend")) 
      .alias("day_type") 
  ])

  return df

In [5]:
from analysis.characterisation.event import event_effect_table

df = base_df(loader=dl, usage_probs=usage_probs).drop(["precip_sum", "wind_max"])
df = df.with_columns(
    temp_ranges(pl.col("temp_max")).alias("temp_range")
)

temp_table = event_effect_table(
    df,
    range_col="temp_range",
    group_cols=("station", "usage_type", "day_type"),
    agg_cols=("usage_type", "day_type"),
)

temp_table

shape: (6, 8)
┌────────────┬──────────┬────────────┬────────────┬────────────┬───────────┬───────────┬───────────┐
│ usage_type ┆ day_type ┆ mean_count ┆ mean_count ┆ mean_count ┆ rel_diff_ ┆ rel_diff_ ┆ rel_diff_ │
│ ---        ┆ ---      ┆ _H         ┆ _M         ┆ _L         ┆ H         ┆ M         ┆ L         │
│ str        ┆ str      ┆ ---        ┆ ---        ┆ ---        ┆ ---       ┆ ---       ┆ ---       │
│            ┆          ┆ f64        ┆ f64        ┆ f64        ┆ f64       ┆ f64       ┆ f64       │
╞════════════╪══════════╪════════════╪════════════╪════════════╪═══════════╪═══════════╪═══════════╡
│ mixed      ┆ weekday  ┆ 2173.38191 ┆ 1599.09243 ┆ 1383.36170 ┆ 69.64     ┆ 21.12     ┆ 0.0       │
│            ┆          ┆            ┆ 7          ┆ 2          ┆           ┆           ┆           │
│ mixed      ┆ weekend  ┆ 1708.45847 ┆ 1096.88135 ┆ 886.703704 ┆ 115.63    ┆ 26.78     ┆ 0.0       │
│            ┆          ┆ 2          ┆ 6          ┆            ┆           ┆           ┆           │
│ recreation ┆ weekday  ┆ 1600.07815 ┆ 1150.51253 ┆ 1009.10050 ┆ 35.89     ┆ 6.88      ┆ 0.0       │
│ al         ┆          ┆ 3          ┆ 5          ┆ 3          ┆           ┆           ┆           │
│ recreation ┆ weekend  ┆ 1375.20990 ┆ 971.041667 ┆ 754.486842 ┆ 63.86     ┆ 16.24     ┆ 0.0       │
│ al         ┆          ┆ 6          ┆            ┆            ┆           ┆           ┆           │
│ utilitaria ┆ weekday  ┆ 5020.91873 ┆ 3895.13970 ┆ 3424.30434 ┆ 38.65     ┆ 10.55     ┆ 0.0       │
│ n          ┆          ┆ 6          ┆ 6          ┆ 8          ┆           ┆           ┆           │
│ utilitaria ┆ weekend  ┆ 4061.06547 ┆ 2811.02336 ┆ 2367.88888 ┆ 61.39     ┆ 17.39     ┆ 0.0       │
│ n          ┆          ┆ 6          ┆ 4          ┆ 9          ┆           ┆           ┆           │
└────────────┴──────────┴────────────┴────────────┴────────────┴───────────┴───────────┴───────────┘

We can see mixed stations react the strongest to temperature change. 
Furthermore weekends are more sensitive to temperature change then weekdays.

**Precipitation**

For precipitation, daily exposure was quantified using the summed daily precipitation.

Again we defined three classes:

$$
\begin{aligned}
\text{L (Low):} \quad & 0~\text{mm} \\
\text{M (Medium):} \quad & 0 < P_{\text{day}} \leq 7.1~\text{mm} \\
\text{H (High):} \quad & P_{\text{day}} > 7.1~\text{mm}
\end{aligned}
$$

implemented in `precip_range`.

In [6]:
def precip_ranges(col):
  return (
    pl.when(col == 0)
    .then(pl.lit("L"))
    .when((col > 0) & (col <= 5))
    .then(pl.lit("M"))
    .otherwise(pl.lit("H"))
  )


As before we will calculate

$$
\text{rel\_diff}_X = \frac{\bar{C}_X - \bar{C}_L}{\bar{C}_L} \times 100 
\quad \text{for } X \in \{M, H\}
$$

In [7]:
df = base_df(loader=dl, usage_probs=usage_probs).drop(["temp_max", "wind_max"])
df = df.with_columns(
    precip_ranges(pl.col("precip_sum")).alias("precip_range")
)

precip_table = event_effect_table(
    df,
    range_col="precip_range",
    group_cols=("station", "usage_type", "day_type"),
    agg_cols=("usage_type", "day_type"),
)

precip_table

shape: (6, 8)
┌────────────┬──────────┬────────────┬────────────┬────────────┬───────────┬───────────┬───────────┐
│ usage_type ┆ day_type ┆ mean_count ┆ mean_count ┆ mean_count ┆ rel_diff_ ┆ rel_diff_ ┆ rel_diff_ │
│ ---        ┆ ---      ┆ _M         ┆ _L         ┆ _H         ┆ M         ┆ L         ┆ H         │
│ str        ┆ str      ┆ ---        ┆ ---        ┆ ---        ┆ ---       ┆ ---       ┆ ---       │
│            ┆          ┆ f64        ┆ f64        ┆ f64        ┆ f64       ┆ f64       ┆ f64       │
╞════════════╪══════════╪════════════╪════════════╪════════════╪═══════════╪═══════════╪═══════════╡
│ mixed      ┆ weekday  ┆ 1793.62390 ┆ 2099.74675 ┆ 1418.32539 ┆ -17.2     ┆ 0.0       ┆ -37.09    │
│            ┆          ┆ 7          ┆ 3          ┆ 7          ┆           ┆           ┆           │
│ mixed      ┆ weekend  ┆ 1316.484   ┆ 1586.57786 ┆ 955.565217 ┆ -19.41    ┆ 0.0       ┆ -46.03    │
│            ┆          ┆            ┆ 9          ┆            ┆           ┆           ┆           │
│ recreation ┆ weekday  ┆ 1318.32142 ┆ 1509.68859 ┆ 1051.58730 ┆ -9.22     ┆ 0.0       ┆ -22.38    │
│ al         ┆          ┆ 9          ┆ 6          ┆ 2          ┆           ┆           ┆           │
│ recreation ┆ weekend  ┆ 1129.96242 ┆ 1274.75072 ┆ 827.986577 ┆ -11.83    ┆ 0.0       ┆ -28.94    │
│ al         ┆          ┆ 8          ┆ 5          ┆            ┆           ┆           ┆           │
│ utilitaria ┆ weekday  ┆ 4252.88601 ┆ 4810.272   ┆ 3451.75694 ┆ -8.49     ┆ 0.0       ┆ -23.46    │
│ n          ┆          ┆            ┆            ┆ 4          ┆           ┆           ┆           │
│ utilitaria ┆ weekend  ┆ 3281.97212 ┆ 3822.52014 ┆ 2404.80172 ┆ -13.27    ┆ 0.0       ┆ -28.83    │
│ n          ┆          ┆ 5          ┆ 7          ┆ 4          ┆           ┆           ┆           │
└────────────┴──────────┴────────────┴────────────┴────────────┴───────────┴───────────┴───────────┘

**Wind Speed**

Daily wind conditions were represented by the mean daily wind speed, as wind affects cycling continuously throughout the day.

$$
\begin{aligned}
\text{L (low):} \quad & \max(W_{\text{day}}) < 12~\text{m/s} \\
\text{M (medium):} \quad & 12 \leq \max(W_{\text{day}}) < 16.8~\text{m/s} \\
\text{H (high):} \quad & \max(W_{\text{day}}) \geq 16.8~\text{m/s}
\end{aligned}
$$

implemented in `wind_ranges`.

In [8]:
def wind_ranges(col):
  return (
    pl.when(col < 12).then(pl.lit("L"))
    .when(col < 16.8).then(pl.lit("M"))
    .otherwise(pl.lit("H"))
  )

In [9]:
df = base_df(loader=dl, usage_probs=usage_probs).drop(["precip_sum", "temp_max"])
df = df.with_columns(
    wind_ranges(pl.col("wind_max")).alias("wind_range")
)

wind_table = event_effect_table(
    df,
    range_col="wind_range",
    group_cols=("station", "usage_type", "day_type"),
    agg_cols=("usage_type", "day_type"),
)

wind_table

shape: (6, 8)
┌────────────┬──────────┬────────────┬────────────┬────────────┬───────────┬───────────┬───────────┐
│ usage_type ┆ day_type ┆ mean_count ┆ mean_count ┆ mean_count ┆ rel_diff_ ┆ rel_diff_ ┆ rel_diff_ │
│ ---        ┆ ---      ┆ _H         ┆ _M         ┆ _L         ┆ H         ┆ M         ┆ L         │
│ str        ┆ str      ┆ ---        ┆ ---        ┆ ---        ┆ ---       ┆ ---       ┆ ---       │
│            ┆          ┆ f64        ┆ f64        ┆ f64        ┆ f64       ┆ f64       ┆ f64       │
╞════════════╪══════════╪════════════╪════════════╪════════════╪═══════════╪═══════════╪═══════════╡
│ mixed      ┆ weekday  ┆ 1700.83098 ┆ 1900.69841 ┆ 1986.02489 ┆ -15.48    ┆ -0.44     ┆ 0.0       │
│            ┆          ┆ 6          ┆ 3          ┆ 6          ┆           ┆           ┆           │
│ mixed      ┆ weekend  ┆ 1228.29383 ┆ 1482.79703 ┆ 1418.86127 ┆ -20.22    ┆ 0.91      ┆ 0.0       │
│            ┆          ┆ 9          ┆            ┆ 2          ┆           ┆           ┆           │
│ recreation ┆ weekday  ┆ 1219.45060 ┆ 1433.99717 ┆ 1423.24431 ┆ -11.56    ┆ -0.39     ┆ 0.0       │
│ al         ┆          ┆ 2          ┆ 5          ┆ 8          ┆           ┆           ┆           │
│ recreation ┆ weekend  ┆ 1045.10793 ┆ 1206.47404 ┆ 1170.53389 ┆ -10.82    ┆ 3.07      ┆ 0.0       │
│ al         ┆          ┆ 7          ┆ 8          ┆ 8          ┆           ┆           ┆           │
│ utilitaria ┆ weekday  ┆ 3983.82131 ┆ 4497.60824 ┆ 4616.51903 ┆ -13.48    ┆ -1.73     ┆ 0.0       │
│ n          ┆          ┆ 7          ┆ 7          ┆ 1          ┆           ┆           ┆           │
│ utilitaria ┆ weekend  ┆ 2990.89156 ┆ 3515.10788 ┆ 3615.90322 ┆ -17.16    ┆ 1.04      ┆ 0.0       │
│ n          ┆          ┆ 6          ┆ 4          ┆ 6          ┆           ┆           ┆           │
└────────────┴──────────┴────────────┴────────────┴────────────┴───────────┴───────────┴───────────┘